File: typescript/core/src/sh.ts
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/**
 * Spherical harmonic bands.
 *
 * A file declares one SH degree for the whole scene, and each band above the constant
 * term is stored in its own record with its own byte range in the chunk index. A reader
 * that has decided to evaluate fewer bands never transfers the ones it will not use.
 *
 * Degrees are whole: band `b` carries every coefficient of that degree for all three
 * colour components, so a reader takes bands 1..D and has exactly a degree-D scene. There
 * is no half a degree, and a reader must not synthesize one by taking part of a band.
 */

import { MalformedFile } from "./errors.js";

/** The highest degree version 1 defines. */
export const MAX_SH_DEGREE = 3;

/**
 * Coefficients per colour component in band `b`: 3, 5, 7 for bands 1, 2, 3.
 *
 * Degree `d` adds `2d + 1` coefficients, which is what makes a degree whole.
 */
export function coefficientsInBand(band: number): number {
  if (band < 1 || band > MAX_SH_DEGREE) {
    throw new MalformedFile(`SH band ${band} is outside the 1..${MAX_SH_DEGREE} range`);
  }
  return 2 * band + 1;
}

/** Coefficients per colour component for a whole degree: 0, 3, 8, 15. */
export function coefficientsForDegree(degree: number): number {
  if (degree < 0 || degree > MAX_SH_DEGREE) {
    throw new MalformedFile(`SH degree ${degree} is outside the 0..${MAX_SH_DEGREE} range`);
  }
  return (degree + 1) * (degree + 1) - 1;
}

/** The half-open coefficient range band `b` occupies within a component's coefficients. */
export function bandCoefficientRange(band: number): readonly [number, number] {
  const first = coefficientsForDegree(band - 1);
  return [first, first + coefficientsInBand(band)];
}

/**
 * Spherical harmonic coefficients for one chunk, component-major.
 *
 * `values[c * coefficients + k]` for gaussian `i` lives at
 * `i * 3 * coefficients + c * coefficients + k`, which is the layout the streams are
 * written in: all of red's coefficients, then green's, then blue's.
 */
export interface ShCoefficients {
  /** Highest whole degree present, after any cap the reader asked for. */
  readonly degree: number;
  /** Coefficients per colour component. */
  readonly coefficients: number;
  readonly count: number;
  readonly values: Uint8Array;
  /** Which bands were actually transferred and merged. */
  readonly bands: readonly number[];
}

/**
 * Merge decoded band streams into one component-major coefficient array.
 *
 * Each band stream is `3 × coefficientsInBand(band)` channels per gaussian, ordered
 * component-major within the band. Bands are placed at their own offset inside the whole
 * degree's coefficient block, so merging never has to know which bands came before it.
 */
export function mergeBands(
  count: number,
  bands: ReadonlyMap<number, Int32Array>,
  degreeCap: number,
): ShCoefficients {
  const present = [...bands.keys()].filter((b) => b <= degreeCap).sort((a, b) => a - b);
  if (present.length === 0) {
    return { degree: 0, coefficients: 0, count, values: new Uint8Array(0), bands: [] };
  }
  // Whole degrees only: bands must run 1..N with nothing missing, or the result would be
  // a scene that is neither degree N nor degree N-1.
  for (let i = 0; i < present.length; i++) {
    if (present[i] !== i + 1) {
      throw new MalformedFile(
        `SH bands ${present.join(", ")} do not form whole degrees starting at band 1`,
      );
    }
  }
  const degree = present.length;
  const coefficients = coefficientsForDegree(degree);
  const values = new Uint8Array(count * 3 * coefficients);
  for (const band of present) {
    const [first] = bandCoefficientRange(band);
    const width = coefficientsInBand(band);
    const stream = bands.get(band)!;
    if (stream.length !== count * 3 * width) {
      throw new MalformedFile(
        `SH band ${band} decoded ${stream.length} values, expected ${count * 3 * width}`,
      );
    }
    for (let i = 0; i < count; i++) {
      for (let c = 0; c < 3; c++) {
        const from = i * 3 * width + c * width;
        const to = i * 3 * coefficients + c * coefficients + first;
        for (let k = 0; k < width; k++) {
          const value = stream[from + k]!;
          if (value < 0 || value > 255) {
            throw new MalformedFile(
              `SH band ${band} coefficient ${value} is outside the 0..255 range this version stores`,
            );
          }
          values[to + k] = value;
        }
      }
    }
  }
  return { degree, coefficients, count, values, bands: present };
}

File: rust/fourdgs/src/sh.rs
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

//! Spherical harmonic bands.
//!
//! A scene declares one degree; each nonconstant band has its own SH Band Stream.
//! Readers take bands whole: 1..D gives exactly a degree-D scene.
//!
//! Coefficients are `u8`, stored as written and consumed as read. `step_sh` in the
//! Quantization record describes what the encoder did before it stored them and is **not**
//! applied at decode — multiplying by it scales appearance by a factor of one to three and
//! is the single most likely way to misread that record (spec §6.5).

use std::collections::BTreeMap;

use crate::error::{Error, Result};
use crate::serialization::MAX_STREAM_BYTES;
use crate::stream::DecodedStream;

/// Coefficients per colour component that each band carries, as `[first, last)` within a
/// whole degree's coefficients. Band `b` carries `2b + 1` of them.
pub fn band_range(band: u8) -> Option<(usize, usize)> {
    match band {
        1 => Some((0, 3)),
        2 => Some((3, 8)),
        3 => Some((8, 15)),
        _ => None,
    }
}

/// Merge per-chunk band streams into one scene-wide `count × (3 * coefficients)` array.
///
/// Columns are component-major — every coefficient of red, then green, then blue — which
/// is the layout the streams are written in.
pub fn merge_chunk_bands(
    counts: &[usize],
    chunk_bands: &[BTreeMap<u8, DecodedStream>],
) -> Result<Option<(Vec<u8>, usize)>> {
    let Some(coefficients) = validate_chunk_bands(counts, chunk_bands)? else {
        return Ok(None);
    };
    let total = counts.iter().try_fold(0usize, |total, count| {
        total.checked_add(*count).ok_or_else(|| {
            Error::UnsupportedOperation("SH gaussian count overflows this platform".into())
        })
    })?;
    let output_length = total
        .checked_mul(3)
        .and_then(|value| value.checked_mul(coefficients))
        .ok_or_else(|| Error::UnsupportedOperation("SH output length overflows".into()))?;
    if output_length > MAX_STREAM_BYTES as usize {
        return Err(Error::UnsupportedOperation(format!(
            "merged SH output needs {output_length} bytes, past the {MAX_STREAM_BYTES} byte decoded-state ceiling"
        )));
    }
    let mut out = vec![0u8; output_length];

    let mut at = 0usize;
    for (chunk, bands) in counts.iter().zip(chunk_bands.iter()) {
        let count = *chunk;
        for (band, stream) in bands {
            let (first, last) = band_range(*band)
                .ok_or_else(|| Error::Malformed(format!("SH band {band} is not 1, 2 or 3")))?;
            let width = last - first;
            for i in 0..count {
                for c in 0..3 {
                    for j in 0..width {
                        let value = stream.get(i, c * width + j);
                        if !(0..=255).contains(&value) {
                            return Err(Error::Malformed(
                                "an SH coefficient is outside the 0..255 range this version stores"
                                    .into(),
                            ));
                        }
                        out[(at + i) * 3 * coefficients + c * coefficients + first + j] =
                            value as u8;
                    }
                }
            }
        }
        at += count;
    }
    Ok(Some((out, coefficients)))
}

/// Validate the logical SH layout without allocating its scene-wide output.
pub(crate) fn validate_chunk_bands(
    counts: &[usize],
    chunk_bands: &[BTreeMap<u8, DecodedStream>],
) -> Result<Option<usize>> {
    if counts.len() != chunk_bands.len() {
        return Err(Error::Malformed(format!(
            "{} chunk counts accompany {} SH band maps",
            counts.len(),
            chunk_bands.len()
        )));
    }
    let mut present: Vec<u8> = Vec::new();
    for bands in chunk_bands {
        for band in bands.keys() {
            if !present.contains(band) {
                present.push(*band);
            }
        }
    }
    present.sort_unstable();
    if present.is_empty() {
        return Ok(None);
    }
    let expected: Vec<u8> = (1..=present.len() as u8).collect();
    if present != expected {
        return Err(Error::Malformed(format!(
            "SH bands {present:?} do not form whole degrees starting at band 1"
        )));
    }

    let coefficients = band_range(*present.last().expect("non-empty"))
        .ok_or_else(|| Error::Malformed(format!("SH band {:?} is not 1, 2 or 3", present.last())))?
        .1;
    for (chunk, bands) in counts.iter().zip(chunk_bands.iter()) {
        let count = *chunk;
        for band in &present {
            let stream = bands.get(band).ok_or_else(|| {
                let mut have: Vec<u8> = bands.keys().copied().collect();
                have.sort_unstable();
                Error::Malformed(format!(
                    "a chunk carries SH bands {have:?}, the file carries {present:?}"
                ))
            })?;
            let (first, last) = band_range(*band)
                .ok_or_else(|| Error::Malformed(format!("SH band {band} is not 1, 2 or 3")))?;
            let width = last - first;
            if stream.channels != 3 * width || stream.count != count {
                return Err(Error::Malformed(format!(
                    "SH band {band} carries {} elements of {} channels; the chunk has {count} gaussians and the band is {} wide",
                    stream.count, stream.channels, 3 * width
                )));
            }
        }
    }
    Ok(Some(coefficients))
}
